### PR TITLE
V0.2 Logical operator overloading for EarlyStoppingChecker

### PR DESCRIPTION
**Pull request by ysqyang**
Logical operator overloading for EarlyStoppingChecker

# Description

Logical operator overloading for EarlyStoppingChecker. 

## Linked issue(s)/Pull request(s)

- [Issue 171](https://github.com/microsoft/maro/issues/171)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [x] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [x] Linux
- Python version:
  - [ ] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [x] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl/early_stopping/abs_early_stopping_checker.py [hunk 1]
```diff
@@ -24,3 +24,66 @@ def __call__(self, metric_series) -> bool:
...
+        """
+        class NotChecker(AbsEarlyStoppingChecker):
+            def __init__(self, checker):
+                super().__init__()
+                self._checker = checker
+
+            def __call__(self, metric_series) -> bool:
+                return not self._checker(metric_series)
+
+        return NotChecker(self)
```
Comment (ArthurJiang) [inline maro/rl/early_stopping/abs_early_stopping_checker.py:89]: InvertChecker

File: examples/cim/dqn/config.yml [hunk 2]
```diff
@@ -7,7 +7,8 @@ general:
   early_stopping:
     warmup_ep: 50
     last_k: 10
-    threshold: 0.05
+    delta_threshold: 0.05
```
Comment (ArthurJiang) [inline examples/cim/dqn/config.yml:10]: add comments for the parameters of early_stopping

File: examples/cim/dqn/single_process_launcher.py [hunk 3]
```diff
@@ -48,10 +49,19 @@ def launch(config):
     )
 
     # Step 4: Create an actor and a learner to start the training process.
-    early_stopping_checker = MaxDeltaEarlyStoppingChecker(
+    early_stopping_checker_1 = SimpleEarlyStoppingChecker(
```
Comment (ArthurJiang) [inline examples/cim/dqn/single_process_launcher.py:52]: perf_checker

File: examples/cim/dqn/single_process_launcher.py [hunk 4]
```diff
@@ -48,10 +49,19 @@ def launch(config):
...
         last_k=config.general.early_stopping.last_k,
-        threshold=config.general.early_stopping.threshold
+        threshold=config.general.early_stopping.perf_threshold,
+        measure_func=lambda vals: mean(vals)
     )
+
+    early_stopping_checker_2 = MaxDeltaEarlyStoppingChecker(
```
Comment (ArthurJiang) [inline examples/cim/dqn/single_process_launcher.py:58]: perf_delta_ehecker

File: examples/cim/dqn/single_process_launcher.py [hunk 5]
```diff
@@ -48,10 +49,19 @@ def launch(config):
...
+        last_k=config.general.early_stopping.last_k,
+        threshold=config.general.early_stopping.delta_threshold
+    )
+
+    early_stopping_checker = early_stopping_checker_1 & early_stopping_checker_2
```
Comment (ArthurJiang) [inline examples/cim/dqn/single_process_launcher.py:63]: combined_checker

File: maro/rl/early_stopping/abs_early_stopping_checker.py [hunk 6]
```diff
@@ -7,10 +7,18 @@
...
+            whether early-stopping should be triggered.
     """
-    def __init__(self):
-        pass
+    def __init__(self, last_k, threshold):
+        super().__init__()
+        self._last_k = last_k
+        self._threshold = threshold
+
+    def is_valid(self, metric_series):
```
Comment (ArthurJiang) [inline maro/rl/early_stopping/abs_early_stopping_checker.py:20]: is_skip?

File: maro/rl/early_stopping/simple_early_stopping_checker.py [hunk 7]
```diff
@@ -2,50 +2,41 @@
...
 
 from .abs_early_stopping_checker import AbsEarlyStoppingChecker
 
 
-class RSDEarlyStoppingChecker(AbsEarlyStoppingChecker):
-    """Early stopping checker based on the mean and standard deviation of the last k metric values.
+class SimpleEarlyStoppingChecker(AbsEarlyStoppingChecker):
+    """Early stopping checker based on the some simple measure obtained by applying a function to the last k metric
+    values."""
+    def __init__(self, last_k, threshold, measure_func: Callable):
```
Comment (ArthurJiang) [inline maro/rl/early_stopping/simple_early_stopping_checker.py:13]: measure_func: Callable[List, Union[int|float]] update thershold:  Union[int|float]

**Review by ArthurJiang — DISMISSED** [review]

File: examples/cim/dqn/single_process_launcher.py (hunk 4 above)
Comment (ysqyang) [reply examples/cim/dqn/single_process_launcher.py:58]: using perf_stability_checker

File: examples/cim/dqn/single_process_launcher.py (hunk 3 above)
Comment (ysqyang) [reply examples/cim/dqn/single_process_launcher.py:52]: using perf_checker

File: maro/rl/early_stopping/abs_early_stopping_checker.py (hunk 6 above)
Comment (ysqyang) [reply maro/rl/early_stopping/abs_early_stopping_checker.py:20]: is_triggered should be better

**Review by ArthurJiang — APPROVED** [review]

**Pull request merged**